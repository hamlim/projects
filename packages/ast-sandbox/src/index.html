<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <meta http-equiv="X-UA-Compatible" content="ie=edge" />
    <title>ast-sandbox</title>
    <style>
      body {
        margin: 0;
        padding: 0;
      }
    </style>
  </head>
  <body>
    <div id="root"></div>
    <script src="./index.js"></script>
    <script src="https://unpkg.com/@babel/standalone@7.9.5/babel.min.js"></script>
  </body>
</html>
