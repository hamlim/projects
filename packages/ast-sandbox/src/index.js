import * as React from 'react'
import {
  BrowserRouter,
  useRoute,
  Link as RerouteLink,
} from '@matthamlin/reroute-browser'
import {
  ThemeProvider,
  H1,
  Box,
  Text,
  Link,
} from '@matthamlin/component-library'
import { createRoot } from 'react-dom'

let { lazy, Suspense } = React

function Route({ path, children }) {
  let props = useRoute(path)
  if (props.match) {
    return children
  }
  return null
}

function RouterLink({ to, children }) {
  return (
    <Link forwardedAs={RerouteLink} to={to}>
      {children}
    </Link>
  )
}

let Sandbox = lazy(() => import('./Sandbox.js'))

function Home() {
  return (
    <Box>
      <H1>AST Sandbox</H1>

      <Text>
        This site is meant to be a sandbox for prototyping Babel plugins and
        other exploratory ast traversal systems.
      </Text>
      <RouterLink to="/sandbox">Go To Sandbox</RouterLink>
    </Box>
  )
}

createRoot(document.querySelector('#root')).render(
  <ThemeProvider>
    <BrowserRouter>
      <Route path="/">
        <Home />
      </Route>
      <Route path="/sandbox">
        <Suspense fallback={<Text>Loading...</Text>}>
          <Sandbox />
        </Suspense>
      </Route>
    </BrowserRouter>
  </ThemeProvider>,
)
