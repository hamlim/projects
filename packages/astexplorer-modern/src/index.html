<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <meta http-equiv="X-UA-Compatible" content="ie=edge" />
    <title>astexplorer-modern</title>
  </head>
  <body>
    <div
      id="root"
      style="display: flex; min-height: 100vh; min-width: 100vw; height: 100vh; width: 100vw; max-height: 100vh; max-width: 100vw; overflow: hidden;"
    ></div>
    <script src="./index.js"></script>
  </body>
</html>
