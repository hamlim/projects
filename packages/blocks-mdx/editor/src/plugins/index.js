// default built-in plugins
import ListsPlugin from '@convertkit/slate-lists'
import DeepTable from 'slate-deep-table'
import SoftBreak from 'slate-soft-break'

import ThemeEditorPlugin from '../plugins/theme-editor'

import MarkdownPlugin from './markdown'
import CodePlugin from './code'
import LiveJSXPlugin from './live-jsx'
import JSXBlocksPlugin from './jsx-blocks'
import TablePlugin from './table'
import ImagePlugin from './image'
import LinkPlugin from './link'
import MarkdownShortcutsPlugin from './markdown-shortcuts'

export default [
  ThemeEditorPlugin(),
  SoftBreak({ shift: true }),
  MarkdownPlugin(),
  CodePlugin(),
  JSXBlocksPlugin(),
  LiveJSXPlugin(),
  TablePlugin(),
  DeepTable(),
  ImagePlugin(),
  LinkPlugin(),
  ListsPlugin({
    blocks: {
      ordered_list: 'numbered-list',
      unordered_list: 'bulleted-list',
      list_item: 'list-item'
    }
  }),
  MarkdownShortcutsPlugin()
]
