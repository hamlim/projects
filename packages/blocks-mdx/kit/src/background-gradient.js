/** @jsx jsx */
import { jsx } from '@emotion/core'
import adaptiveGradient from 'adaptive-gradient'
import randomHex from 'random-hex-color'

const BackgroundGradient = ({
  start = 'rebeccapurple',
  end,
  stop = '80%',
  angle = 30,
  random,
  ...props
}) => {
  if (random) {
    start = randomHex()
  }

  const gradient = adaptiveGradient(start)
  const gradientEnd = '#' + gradient.end
  const backgroundGradient = `
    linear-gradient(
      ${angle}deg,
      ${start},
      ${end || gradientEnd} ${stop}
    )
  `

  return (
    <div
      css={{
        backgroundColor: start,
        background: backgroundGradient,
        color: gradient.fontShouldBeLight ? 'white' : 'black',
        padding: 20
      }}
      {...props}
    />
  )
}

BackgroundGradient.propertyControls = {
  isVoid: false,
  start: {
    type: 'color',
    description: 'Starting color'
  },
  end: {
    type: 'color',
    description: 'Ending color'
  },
  stop: {
    type: 'string',
    description: 'Blending stopping point'
  },
  angle: {
    type: 'string',
    description: 'Angle of the gradient'
  },
  random: {
    type: 'boolean',
    description: 'Randomize the gradient'
  }
}

export default BackgroundGradient
