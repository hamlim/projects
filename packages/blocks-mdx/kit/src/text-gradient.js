/** @jsx jsx */
import { jsx } from '@emotion/core'
import adaptiveGradient from 'adaptive-gradient'
import randomHex from 'random-hex-color'

const TextGradient = ({
  start = 'rebeccapurple',
  end,
  stop = '60%',
  angle = 90,
  random,
  ...props
}) => {
  if (random) {
    start = randomHex()
  }

  const gradient = adaptiveGradient(start)
  const gradientEnd = '#' + gradient.end
  const backgroundGradient = `
    linear-gradient(
      ${angle}deg,
      ${start},
      ${end || gradientEnd} ${stop}
    )
  `

  return (
    <div
      css={{
        backgroundColor: start,
        background: backgroundGradient,
        WebkitBackgroundClip: 'text',
        WebkitTextFillColor: 'transparent'
      }}
      {...props}
    />
  )
}

TextGradient.propertyControls = {
  isVoid: false,
  start: {
    type: 'color'
  },
  end: {
    type: 'color'
  },
  stop: {
    type: 'string'
  },
  angle: {
    type: 'string'
  },
  random: {
    type: 'boolean'
  }
}

export default TextGradient
