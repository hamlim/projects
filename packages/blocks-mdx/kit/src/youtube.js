/** @jsx jsx */
import { jsx } from '@emotion/core'

import Player from 'react-youtube'
import isURL from 'is-url'
import getYouTubeID from 'get-youtube-id'

const formatValue = n => (isURL(n) ? getYouTubeID(n) : n)

const Wrapper = props => (
  <div
    {...props}
    css={{
      width: '100%',
      height: 0,
      paddingBottom: 900 / 16 + '%',
      position: 'relative',
      overflow: 'hidden',
      backgroundColor: '#eee',
      iframe: {
        position: 'absolute',
        width: '100%',
        height: '100%',
        top: 0,
        bottom: 0,
        left: 0,
        border: 0
      }
    }}
  />
)

const YouTube = ({ videoId, ...props }) => {
  return (
    <Wrapper>
      <Player videoId={formatValue(videoId)} {...props} />
    </Wrapper>
  )
}

// mimicking Framer X - could make this compatible
YouTube.propertyControls = {
  isVoid: true,
  videoId: {
    type: 'string',
    title: 'Video ID',
    description: 'YouTube ID of the video',
    formatValue
  }
}

export default YouTube
