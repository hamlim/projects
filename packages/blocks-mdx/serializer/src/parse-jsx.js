// utilities for parsing jsx block strings from MDX into slate schema
const babel = require('@babel/core')
const { declare } = require('@babel/helper-plugin-utils')
const plugins = [require('@babel/plugin-syntax-jsx')]

class PluginGetRootElement {
  constructor() {
    const result = {
      props: {}
    }
    this.result = result

    this.plugin = declare(babel => {
      babel.assertVersion(7)

      const visitProps = {
        JSXAttribute(path) {
          const key = path.node.name.name
          // only handles string/static props
          const value = path.node.value.value
          result.props[key] = value
        }
      }

      return {
        visitor: {
          JSXOpeningElement(path) {
            // only parse root-level element
            if (result.type) return
            result.type = path.node.name.name
            path.traverse(visitProps)
          }
        }
      }
    })
  }
}

module.exports.parseJSXBlock = jsx => {
  const { plugin, result } = new PluginGetRootElement()
  try {
    babel.transformSync(jsx, {
      plugins: [...plugins, plugin]
    })
    const { type, props } = result
    return {
      type,
      props
    }
  } catch (e) {
    return {
      props: {}
    }
  }
}

function pluginApplyProps(babel, state) {
  let applied = false
  const { types: t } = babel

  const addProps = (path, opts) => {
    const { props = {} } = opts
    for (const key in props) {
      const value = props[key]
      const id = t.jSXIdentifier(key)
      const attribute = t.jSXAttribute(id, t.stringLiteral(value))
      const index = path.node.attributes.findIndex(
        attr => attr.name.name === key
      )
      if (index < 0) {
        path.node.attributes.push(attribute)
      } else {
        path.node.attributes[index] = attribute
      }
    }
  }

  const visitor = {
    JSXOpeningElement(path, state) {
      // only apply props to root element
      if (applied) return
      addProps(path, state.opts)
      applied = true
    }
  }

  return {
    visitor
  }
}

module.exports.applyProps = (jsx, opts) => {
  try {
    const result = babel.transformSync(jsx, {
      plugins: [...plugins, [pluginApplyProps, opts]]
    })
    const next = result.code.replace(/;$/, '')
    return next
  } catch (e) {
    return jsx
  }
}
