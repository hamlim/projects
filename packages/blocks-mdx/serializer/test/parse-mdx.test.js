import { parseMDX } from '../src'

const FIXTURE = `
# Hello, __world!__

<Block>

## Other

Stuff

</Block>

And more stuff
`

test('parses a basic block', () => {
  const result = parseMDX('# Hello, world!')

  expect(result).toMatchSnapshot()
})

test('handles an interleaved block', () => {
  const result = parseMDX(FIXTURE)

  expect(result).toMatchSnapshot()
})
