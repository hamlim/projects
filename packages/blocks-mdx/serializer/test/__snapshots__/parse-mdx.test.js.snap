// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handles an interleaved block 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "type": "text",
          "value": "Hello, ",
        },
        Object {
          "children": Array [
            Object {
              "type": "text",
              "value": "world!",
            },
          ],
          "type": "strong",
        },
      ],
      "depth": 1,
      "type": "heading",
    },
    Object {
      "children": Array [
        Object {
          "type": "jsx",
          "value": "<Block>",
        },
        Object {
          "children": Array [
            Object {
              "type": "text",
              "value": "Other",
            },
          ],
          "depth": 2,
          "type": "heading",
        },
        Object {
          "children": Array [
            Object {
              "type": "text",
              "value": "Stuff",
            },
          ],
          "type": "paragraph",
        },
        Object {
          "type": "jsx",
          "value": "</Block>",
        },
      ],
      "type": "jsx",
    },
    Object {
      "children": Array [
        Object {
          "type": "text",
          "value": "And more stuff",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "root",
}
`;

exports[`parses a basic block 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "type": "text",
          "value": "Hello, world!",
        },
      ],
      "depth": 1,
      "type": "heading",
    },
  ],
  "type": "root",
}
`;
